src/modules/rover_ackermann/RoverAckermannGuidance/RoverAckermannGuidance.cpp

```diff
@@ -0,0 +1,319 @@
+		acceptance_radius = math::constrain<float>(acceptance_radius_temp_scaled, _param_ra_acc_rad_def.get(),
+				    _param_ra_acc_rad_max.get());
+	}
+
+	// Publish updated acceptance radius
+	position_controller_status_s pos_ctrl_status = {};
+	pos_ctrl_status.acceptance_radius = acceptance_radius;
+	pos_ctrl_status.timestamp = hrt_absolute_time();
+	_pos_ctrl_status_pub.publish(pos_ctrl_status);
+}
+
+float RoverAckermannGuidance::calcDesiredSteering()
+{
+	// Calculate crosstrack error
+	const Vector2f prev_wp_to_curr_wp_local = _curr_wp_local - _prev_wp_local;
+
+	if (prev_wp_to_curr_wp_local.norm() < FLT_EPSILON) { // Avoid division by 0 (this case should not happen)
+		return 0.f;
+	}
+
+	const Vector2f prev_wp_to_curr_pos_local = _curr_pos_local - _prev_wp_local;
+	const Vector2f distance_on_line_segment = ((prev_wp_to_curr_pos_local * prev_wp_to_curr_wp_local) /
+			prev_wp_to_curr_wp_local.norm()) * prev_wp_to_curr_wp_local.normalized();
+	const Vector2f crosstrack_error = (_prev_wp_local + distance_on_line_segment) - _curr_pos_local;
+	_rover_ackermann_guidance_status.crosstrack_error = crosstrack_error.norm(); // For logging
+
+	// Calculate desired heading towards lookahead point
+	float desired_heading{0.f};
+	float lookahead_distance = math::constrain(_param_ra_lookahead_tun.get() * _desired_speed,
+				   _param_ra_lookahead_min.get(), _param_ra_lookahead_max.get());
+
+	if (crosstrack_error.longerThan(lookahead_distance)) {
+		if (crosstrack_error.norm() < _param_ra_lookahead_max.get()) {
+			lookahead_distance = crosstrack_error.norm(); // Scale lookahead radius
+			desired_heading = calcDesiredHeading(_curr_wp_local, _prev_wp_local, _curr_pos_local, lookahead_distance);
+
+		} else { // Excessively large crosstrack error
+			desired_heading = calcDesiredHeading(_curr_wp_local, _curr_pos_local, _curr_pos_local, lookahead_distance);
+		}
+
+	} else { // Crosstrack error smaller than lookahead
+		desired_heading = calcDesiredHeading(_curr_wp_local, _prev_wp_local, _curr_pos_local, lookahead_distance);
+	}
+
+	// Calculate desired steering to reach lookahead point
+	const float heading_error = matrix::wrap_pi(desired_heading - _vehicle_yaw);
+	_rover_ackermann_guidance_status.lookahead_distance = lookahead_distance; // For logging
+	_rover_ackermann_guidance_status.heading_error = (heading_error * 180.f) / (M_PI_F); // For logging
+
+	if (math::abs_t(heading_error) <= M_PI_2_F) {
+		return atanf(2 * _param_ra_wheel_base.get() * sinf(heading_error) / lookahead_distance);
+
+	} else if (heading_error > FLT_EPSILON) {
+		return atanf(2 * _param_ra_wheel_base.get() * (1.0f + sinf(heading_error - M_PI_2_F)) /
+			     lookahead_distance);
+
+	} else {
+		return atanf(2 * _param_ra_wheel_base.get() * (-1.0f + sinf(heading_error + M_PI_2_F)) /
+			     lookahead_distance);
+	}
+}
+
+float RoverAckermannGuidance::calcDesiredHeading(const Vector2f &curr_wp_local, const Vector2f &prev_wp_local,
```

Comment:
Would be cool to have unit tests for this (and similarly complex) functions in here. That would be a way to check that it produces the right outputs for the whole input range.

Reply:
I agree but i would propose the following: Since the pure pursuit controller might also be suitable for differential rovers and even boats (@PerFrivik is working on that) it would make sense to eventually turn it into a library like the old l1 guidance.